File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/doc-content.ts
#!/usr/bin/env bun
import type {
	Finding,
	HarnessCheckRule,
	HarnessManifest,
	RuleContext,
} from "../harness-check-rule";

type ContentCondition =
	| string
	| readonly unknown[]
	| {
			readonly contains?: unknown;
			readonly allOf?: unknown;
			readonly anyOf?: unknown;
			readonly not?: unknown;
		};

/**
 * Require specified content in documentation files.
 */
export const docContentRule = (ctx: RuleContext): HarnessCheckRule => ({
	category: "docContent",
	applies(manifest: HarnessManifest): boolean {
		const section = manifest.docContent;
		if (typeof section !== "object" || section === null) {
			return false;
		}
		const enabled = (section as { enabled?: unknown }).enabled;
		if (enabled === false) {
			return false;
		}
		const checks = ctx.readJsonObject(
			(section as Record<string, unknown>).parameters,
		).checks;
		return Array.isArray(checks) && checks.length > 0;
	},

	validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
		const checks = ctx.readJsonObject(
			ctx.readJsonObject(manifest.docContent).parameters,
		).checks;
		if (!Array.isArray(checks)) {
			return [];
		}
		return checks
			.filter((check) => typeof check === "object" && check !== null)
			.flatMap((check) => {
				const checkObj = check as Record<string, unknown>;
				const files = ctx.readStringArray(checkObj.files);
				const failureMessage =
					typeof checkObj.failureMessage === "string"
						? checkObj.failureMessage
						: "";
				const combinedText = files.map((f) => ctx.read(f)).join("\n");
				return !conditionMatches(checkObj, combinedText) && failureMessage
					? [
							{
								severity: ctx.severityOf(manifest, "docContent"),
								category: "docContent",
								message: failureMessage,
							},
						]
					: [];
			});
	},
});

const conditionMatches = (
	checkObj: Record<string, unknown>,
	combinedText: string,
): boolean => {
	const condition = checkObj.condition ?? checkObj.when;
	if (condition !== undefined) {
		return evaluateCondition(condition as ContentCondition, combinedText);
	}
	return false;
};

const evaluateCondition = (
	condition: ContentCondition,
	combinedText: string,
): boolean => {
	if (typeof condition === "string") {
		return combinedText.includes(condition);
	}
	if (Array.isArray(condition)) {
		return condition.every((item) =>
			evaluateCondition(item as ContentCondition, combinedText),
		);
	}
	if (typeof condition !== "object" || condition === null) {
		return false;
	}
	const hasAll = "allOf" in condition;
	const hasAny = "anyOf" in condition;
	const hasContains = "contains" in condition;
	const hasNot = "not" in condition;
	if (!(hasAll || hasAny || hasContains || hasNot)) {
		return false;
	}
	const allOf = readConditionArray(condition.allOf);
	const anyOf = readConditionArray(condition.anyOf);
	const contains = readStringArray(condition.contains);
	const notCondition = condition.not;
	const andMatches =
		!hasAll || allOf.every((item) => evaluateCondition(item, combinedText));
	const orMatches =
		!hasAny || anyOf.some((item) => evaluateCondition(item, combinedText));
	const containsMatches = contains.every((substring) =>
		combinedText.includes(substring),
	);
	const notMatches =
		!hasNot || !evaluateCondition(notCondition as ContentCondition, combinedText);
	return andMatches && orMatches && containsMatches && notMatches;
};

const readConditionArray = (value: unknown): readonly ContentCondition[] =>
	Array.isArray(value)
		? value.map((item) => item as ContentCondition)
		: typeof value === "string" || (typeof value === "object" && value !== null)
			? [value as ContentCondition]
			: [];

const readStringArray = (value: unknown): readonly string[] =>
	Array.isArray(value)
		? value.filter((item): item is string => typeof item === "string")
		: typeof value === "string"
			? [value]
			: [];

File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/doc_content.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require doc content rule."""

from collections.abc import Iterable
from pathlib import Path
from typing import override

from harness_check_rule import Finding, HarnessCheckRule, JsonObject

from .utils import is_safe_file, read_text, severity_for


class DocContentRule(HarnessCheckRule):
    """Validate docContent check."""

    category: str = "docContent"

    @override
    def applies(self, manifest: JsonObject) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not HarnessCheckRule.is_json_object(section):
            return False
        return section.get("enabled", True) is not False

    @override
    def validate(self, project_dir: Path, manifest: JsonObject) -> Iterable[Finding]:
        """Validate docContent check."""
        section = manifest.get(self.category, {})
        if not HarnessCheckRule.is_json_object(section):
            return []
        params = section.get("parameters", {})
        if not HarnessCheckRule.is_json_object(params):
            return []
        checks = HarnessCheckRule.json_array(params.get("checks", []))
        findings: list[Finding] = []
        for check in checks:
            if not HarnessCheckRule.is_json_object(check):
                continue
            if not isinstance(check.get("files"), list):
                continue
            failure_message = check.get("failureMessage")
            if not isinstance(failure_message, str):
                continue
            if DocContentRule.condition_matches(check, DocContentRule.combined_text(project_dir, check)):
                continue
            findings.append(Finding(severity_for(manifest, self.category), self.category, failure_message))
        return findings

    @staticmethod
    def combined_text(root: Path, check: JsonObject) -> str:
        """Return newline-joined text from safe files declared by a check."""
        return "\n".join(
            read_text(root / f)
            for f in HarnessCheckRule.json_array(check.get("files", []))
            if isinstance(f, str) and is_safe_file(root / f)
        )

    @staticmethod
    def condition_matches(check: JsonObject, content: str) -> bool:
        """Evaluate a nested condition expression."""
        condition = check.get("condition", check.get("when"))
        if condition is not None:
            return DocContentRule.evaluate_condition(condition, content)
        return False

    @staticmethod
    def evaluate_condition(condition: object, content: str) -> bool:
        """Evaluate nested allOf / anyOf / not / contains content conditions."""
        if isinstance(condition, str):
            return condition in content
        if HarnessCheckRule.is_json_array(condition):
            return all(DocContentRule.evaluate_condition(item, content) for item in condition)
        if not HarnessCheckRule.is_json_object(condition):
            return False
        has_all = "allOf" in condition
        has_any = "anyOf" in condition
        has_contains = "contains" in condition
        has_not = "not" in condition
        if not any((has_all, has_any, has_contains, has_not)):
            return False
        all_of = DocContentRule.condition_array(condition.get("allOf"))
        any_of = DocContentRule.condition_array(condition.get("anyOf"))
        contains = DocContentRule.string_array(condition.get("contains"))
        and_matches = not has_all or all(DocContentRule.evaluate_condition(item, content) for item in all_of)
        or_matches = not has_any or any(DocContentRule.evaluate_condition(item, content) for item in any_of)
        contains_matches = all(item in content for item in contains)
        not_condition = condition.get("not")
        not_matches = not has_not or not DocContentRule.evaluate_condition(not_condition, content)
        return and_matches and or_matches and contains_matches and not_matches

    @staticmethod
    def condition_array(value: object) -> list[object]:
        """Normalize a condition value into a list of condition nodes."""
        if HarnessCheckRule.is_json_array(value):
            return value
        if isinstance(value, str) or HarnessCheckRule.is_json_object(value):
            return [value]
        return []

    @staticmethod
    def string_array(value: object) -> list[str]:
        """Normalize a string or string array into string items."""
        if isinstance(value, str):
            return [value]
        if HarnessCheckRule.is_json_array(value):
            return [item for item in value if isinstance(item, str)]
        return []


RULE: HarnessCheckRule = DocContentRule()
